### PR TITLE
Allow integrations customization

### PR DESCRIPTION
**Pull request by mateocervilla**
|Related issue|
|---|
|#14479|

## Description

With the aim of allowing users to modify integrations variables, this PR adds a <options> tag to the <integration> block where they can set the variables included on the integrations APIs. This options must be set in JSON format.

## Configuration options

**ossec.conf configuration:**
```
<integration>
    <name>slack</name>
    <hook_url>https://hooks.slack.com/services/XXXXXXXXXXXXXX</hook_url>
    <alert_format>json</alert_format>
    <options>{"pretext": "Custom Title"}</options>
</integration>
```

## Integrations
This modification was introduced to the actual integrations which are:

- Slack
- Shuffle
- Pagerduty

## Tests

We have tested each integration with the configuration described below, occasionally using the options tag.

#### Configuration

```xml
  <integration>
    <name>slack</name>
    <hook_url>https://hooks.slack.com/services/XXXXX/XXXXXXXXXXXXXX</hook_url> <!-- Replace with your Slack hook URL -->
    <alert_format>json</alert_format>
    <options>{"pretext": "Wazuh IntegratorD"}</options><!-- Optional -->
  </integration>
```

<details><summary><h5>Results from ossec.log</h5></summary>

```
2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:439 at OS_IntegratorD(): DEBUG: # Running Slack script

2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening options file at '/tmp/slack-1679596811-1313538660.options' with '{'client': 'Wazuh IntegratorD'}'

2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening alert file at '/tmp/slack-1679596811-249779904.alert' with '{'timestamp': '2023-03-23T15:40:07.959-0300', 'rule': {'level': 7, 'description': 'Host-based anomaly detection event (rootcheck).', 'id': '510', 'firedtimes': 2, 'mail': False, 'groups': ['ossec', 'rootcheck'], 'pci_dss': ['10.6.1'], 'gdpr': ['IV_35.7.d']}, 'agent': {'id': '000', 'name': 'manager'}, 'manager': {'name': 'manager'}, 'id': '1679596807.35800', 'full_log': "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", 'decoder': {'name': 'rootcheck'}, 'data': {'title': 'Trojaned version of file detected.', 'file': '/usr/bin/diff'}, 'location': 'rootcheck'}'

2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:439 at OS_IntegratorD(): DEBUG: # Generating message

2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:439 at OS_IntegratorD(): DEBUG: # Sending message {"attachments": [{"color": "warning", "pretext": "WAZUH Alert", "title": "Host-based anomaly detection event (rootcheck).", "text": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "fields": [{"title": "Agent", "value": "(000) - manager"}, {"title": "Location", "value": "rootcheck"}, {"title": "Rule ID", "value": "510 _(Level 7)_"}], "ts": "1679596807.35800", "client": "Wazuh IntegratorD"}]} to Slack server

2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:439 at OS_IntegratorD(): DEBUG: # Response received: <bound method Response.json of <Response [200]>>

2023/03/23 15:40:12 wazuh-integratord[155601] integrator.c:452 at OS_IntegratorD(): DEBUG: Command ran successfully.

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Running Slack script

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening options file at '/tmp/slack-1679596895--378924943.options' with '{'pretext': 'Wazuh IntegratorD'}'

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening alert file at '/tmp/slack-1679596895--1103396731.alert' with '{'timestamp': '2023-03-23T15:41:31.888-0300', 'rule': {'level': 7, 'description': 'Host-based anomaly detection event (rootcheck).', 'id': '510', 'firedtimes': 2, 'mail': False, 'groups': ['ossec', 'rootcheck'], 'pci_dss': ['10.6.1'], 'gdpr': ['IV_35.7.d']}, 'agent': {'id': '000', 'name': 'manager'}, 'manager': {'name': 'manager'}, 'id': '1679596891.36805', 'full_log': "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", 'decoder': {'name': 'rootcheck'}, 'data': {'title': 'Trojaned version of file detected.', 'file': '/usr/bin/diff'}, 'location': 'rootcheck'}'

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Generating message

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Sending message {"attachments": [{"color": "warning", "pretext": "Wazuh IntegratorD", "title": "Host-based anomaly detection event (rootcheck).", "text": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "fields": [{"title": "Agent", "value": "(000) - manager"}, {"title": "Location", "value": "rootcheck"}, {"title": "Rule ID", "value": "510 _(Level 7)_"}], "ts": "1679596891.36805"}]} to Slack server

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Response received: <bound method Response.json of <Response [200]>>

2023/03/23 15:41:35 wazuh-integratord[157468] integrator.c:452 at OS_IntegratorD(): DEBUG: Command ran successfully.
```
</details>

#### Configuration

```xml
  <integration>
     <name>shuffle</name>
     <level>5</level>
     <hook_url>http://XXX.XXX.XXX.XXX:YYYY/api/v1/hooks/ZZZZZZ</hook_url>
     <alert_format>json</alert_format>
     <options>{"data": {"title": "VERY IMPORTANT !"}}</options>
  </integration> 
```

<details><summary><h5>Results from ossec.log</h5></summary>

```2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # Running Shuffle script

2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # JSON file for options  doesn't exist

2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening options file at '' with 'None'

2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening alert file at '/tmp/shuffle-1679616773--1244500717.alert' with '{'timestamp': '2023-03-23T21:12:51.978-0300', 'rule': {'level': 7, 'description': 'Host-based anomaly detection event (rootcheck).', 'id': '510', 'firedtimes': 1, 'mail': False, 'groups': ['ossec', 'rootcheck'], 'pci_dss': ['10.6.1'], 'gdpr': ['IV_35.7.d']}, 'agent': {'id': '000', 'name': 'manager'}, 'manager': {'name': 'manager'}, 'id': '1679616771.41347', 'full_log': "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", 'decoder': {'name': 'rootcheck'}, 'data': {'title': 'Trojaned version of file detected.', 'file': '/usr/bin/diff'}, 'location': 'rootcheck'}'

2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # Generating message

2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # Sending message {"severity": 2, "pretext": "WAZUH Alert", "title": "Host-based anomaly detection event (rootcheck).", "text": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "rule_id": "510", "timestamp": "2023-03-23T21:12:51.978-0300", "id": "1679616771.41347", "all_fields": {"timestamp": "2023-03-23T21:12:51.978-0300", "rule": {"level": 7, "description": "Host-based anomaly detection event (rootcheck).", "id": "510", "firedtimes": 1, "mail": false, "groups": ["ossec", "rootcheck"], "pci_dss": ["10.6.1"], "gdpr": ["IV_35.7.d"]}, "agent": {"id": "000", "name": "manager"}, "manager": {"name": "manager"}, "id": "1679616771.41347", "full_log": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "decoder": {"name": "rootcheck"}, "data": {"title": "Trojaned version of file detected.", "file": "/usr/bin/diff"}, "location": "rootcheck"}} to Shuffle server

2023/03/23 21:12:53 wazuh-integratord[161677] integrator.c:439 at OS_IntegratorD(): DEBUG: # Response received: {'success': True, 'execution_id': 'aebc53cb-9234-4498-b4f2-516fd8226311'}

2023/03/23 21:17:14 wazuh-integratord[163501] integrator.c:439 at OS_IntegratorD(): DEBUG: # Running Shuffle script

2023/03/23 21:17:14 wazuh-integratord[163501] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening options file at '/tmp/shuffle-1679617034--1074474089.options' with '{'data': {'title': 'VERY IMPORTANT !'}}'

2023/03/23 21:17:14 wazuh-integratord[163501] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening alert file at '/tmp/shuffle-1679617034--901830250.alert' with '{'timestamp': '2023-03-23T21:17:13.527-0300', 'rule': {'level': 7, 'description': 'Host-based anomaly detection event (rootcheck).', 'id': '510', 'firedtimes': 2, 'mail': False, 'groups': ['ossec', 'rootcheck'], 'pci_dss': ['10.6.1'], 'gdpr': ['IV_35.7.d']}, 'agent': {'id': '000', 'name': 'manager'}, 'manager': {'name': 'manager'}, 'id': '1679617033.42727', 'full_log': "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", 'decoder': {'name': 'rootcheck'}, 'data': {'title': 'Trojaned version of file detected.', 'file': '/usr/bin/diff'}, 'location': 'rootcheck'}'

2023/03/23 21:17:14 wazuh-integratord[163501] integrator.c:439 at OS_IntegratorD(): DEBUG: # Generating message

2023/03/23 21:17:14 wazuh-integratord[163501] integrator.c:439 at OS_IntegratorD(): DEBUG: # Sending message {"severity": 2, "pretext": "WAZUH Alert", "title": "Host-based anomaly detection event (rootcheck).", "text": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "rule_id": "510", "timestamp": "2023-03-23T21:17:13.527-0300", "id": "1679617033.42727", "all_fields": {"timestamp": "2023-03-23T21:17:13.527-0300", "rule": {"level": 7, "description": "Host-based anomaly detection event (rootcheck).", "id": "510", "firedtimes": 2, "mail": false, "groups": ["ossec", "rootcheck"], "pci_dss": ["10.6.1"], "gdpr": ["IV_35.7.d"]}, "agent": {"id": "000", "name": "manager"}, "manager": {"name": "manager"}, "id": "1679617033.42727", "full_log": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "decoder": {"name": "rootcheck"}, "data": {"title": "Trojaned version of file detected.", "file": "/usr/bin/diff"}, "location": "rootcheck"}, "data": {"title": "VERY IMPORTANT !"}} to Shuffle server
```
</details>

#### Configuration

```xml
  <integration>
    <name>pagerduty</name>
     <api_key>XXXXXXXXXX</api_key> <!-- Replace with your PagerDuty API key -->
    <alert_format>json</alert_format><!-- With the new script this is mandatory -->
    <options>{"pretext": "Wazuh IntegratorD"}</options><!-- Optional -->
  </integration>
```

<details><summary><h5>Results from ossec.log</h5></summary>

```
2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # Running PagerDuty script

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # JSON file for options  doesn't exist

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening options file at '' with 'None'

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening alert file at '/tmp/pagerduty-1679594376--655618819.alert' with '{'timestamp': '2023-03-23T14:59:34.202-0300', 'rule': {'level': 5, 'description': 'File added to the system.', 'id': '554', 'firedtimes': 1, 'mail': False, 'groups': ['ossec', 'syscheck', 'syscheck_entry_added', 'syscheck_file'], 'pci_dss': ['11.5'], 'gpg13': ['4.11'], 'gdpr': ['II_5.1.f'], 'hipaa': ['164.312.c.1', '164.312.c.2'], 'nist_800_53': ['SI.7'], 'tsc': ['PI1.4', 'PI1.5', 'CC6.1', 'CC6.8', 'CC7.2', 'CC7.3']}, 'agent': {'id': '000', 'name': 'manager'}, 'manager': {'name': 'manager'}, 'id': '1679594374.31287', 'full_log': "File '/home/administrator/Folder/10.txt' added\nMode: scheduled\n", 'syscheck': {'path': '/home/administrator/Folder/10.txt', 'mode': 'scheduled', 'size_after': '0', 'perm_after': 'rw-rw-r--', 'uid_after': '1000', 'gid_after': '1000', 'md5_after': 'd41d8cd98f00b204e9800998ecf8427e', 'sha1_after': 'da39a3ee5e6b4b0d3255bfef95601890afd80709', 'sha256_after': 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855', 'uname_after': 'administrator', 'gname_after': 'administrator', 'mtime_after': '2023-03-23T14:59:33', 'inode_after': 1996874, 'event': 'added'}, 'decoder': {'name': 'syscheck_new_entry'}, 'location': 'syscheck'}'

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # Generating message

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # Sending message

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:439 at OS_IntegratorD(): DEBUG: # Response received: {'dedup_key': '20f6f2f077ed4ebc974a5799b6745b4a', 'message': 'Event processed', 'status': 'success'}

2023/03/23 14:59:38 wazuh-integratord[149737] integrator.c:452 at OS_IntegratorD(): DEBUG: Command ran successfully.

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Running PagerDuty script

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening options file at '/tmp/pagerduty-1679596895-1212495611.options' with '{'client': 'Wazuh IntegradorD'}'

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Opening alert file at '/tmp/pagerduty-1679596895--316116726.alert' with '{'timestamp': '2023-03-23T15:41:31.888-0300', 'rule': {'level': 7, 'description': 'Host-based anomaly detection event (rootcheck).', 'id': '510', 'firedtimes': 2, 'mail': False, 'groups': ['ossec', 'rootcheck'], 'pci_dss': ['10.6.1'], 'gdpr': ['IV_35.7.d']}, 'agent': {'id': '000', 'name': 'manager'}, 'manager': {'name': 'manager'}, 'id': '1679596891.36805', 'full_log': "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", 'decoder': {'name': 'rootcheck'}, 'data': {'title': 'Trojaned version of file detected.', 'file': '/usr/bin/diff'}, 'location': 'rootcheck'}'

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Generating message

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Sending message {"routing_key": "1b7753d129374006c01b9a229d61594d", "event_action": "trigger", "payload": {"summary": "Host-based anomaly detection event (rootcheck).", "timestamp": "2023-03-23T15:41:31.888-0300", "source": "manager", "severity": "warning", "group": "ossec, rootcheck", "custom_details": {"timestamp": "2023-03-23T15:41:31.888-0300", "rule": {"level": 7, "description": "Host-based anomaly detection event (rootcheck).", "id": "510", "firedtimes": 2, "mail": false, "groups": ["ossec", "rootcheck"], "pci_dss": ["10.6.1"], "gdpr": ["IV_35.7.d"]}, "agent": {"id": "000", "name": "manager"}, "manager": {"name": "manager"}, "id": "1679596891.36805", "full_log": "Trojaned version of file '/usr/bin/diff' detected. Signature used: 'bash|^/bin/sh|file\\.h|proc\\.h|/dev/[^n]|^/bin/.*sh' (Generic).", "decoder": {"name": "rootcheck"}, "data": {"title": "Trojaned version of file detected.", "file": "/usr/bin/diff"}, "location": "rootcheck"}}, "client": "Wazuh IntegradorD", "client_url": "https://wazuh.com/"} to PagerDuty server

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:439 at OS_IntegratorD(): DEBUG: # Response received: {'dedup_key': '390d5a7f418841f2b17d9d5b3b46430b', 'message': 'Event processed', 'status': 'success'}

2023/03/23 15:41:37 wazuh-integratord[157468] integrator.c:452 at OS_IntegratorD(): DEBUG: Command ran successfully.
```
</details>

### 🟢  Memory and static code analysis

[Coverage and ScanBuild reports](https://github.com/wazuh/wazuh/files/10449571/Reports.zip)

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [ ] Package installation
- [x] Source upgrade
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [ ] Coverity
  - [ ] Valgrind (memcheck and descriptor leaks check)
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

File: integrations/pagerduty.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            If no valid JSON file are used
+    """
+    try:
+        with open(file_location) as alert_file:
+            return json.load(alert_file)
+    except FileNotFoundError:
+        debug("# JSON file %s doesn't exist" % file_location)
+        sys.exit(ERR_FILE_NOT_FOUND)
+    except json.decoder.JSONDecodeError as e:
+        debug("Failed getting json_alert %s" % e)
```
Comment (MarcelKemp) [inline integrations/pagerduty.py:245]: Modify debug message to be generic.
```suggestion
        debug("Failed getting JSON file. Error: %s" % e)
```
Change in all integrations.

File: integrations/pagerduty.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        Returns
+        -------
+        {}: any
+            The JSON object read it.
+
+        Raises
+        ------
+        FileNotFoundError
+            If no alert file is not present.
```
Comment (MarcelKemp) [inline integrations/pagerduty.py:234]:
```suggestion
            If no JSON file is found.
```

File: integrations/pagerduty.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    """
+
+    headers = {'content-type': 'application/json', 'Accept-Charset': 'UTF-8'}
+    url     = 'https://events.pagerduty.com/v2/enqueue'
+    res     = requests.post(url, data=msg, headers=headers)
+    debug("# Response received: %s" % res)
+
+def get_json_file(file_location: str) -> any:
+    """
+        Read the JSON object from alert file
```
Comment (MarcelKemp) [inline integrations/pagerduty.py:219]:
```suggestion
        Read JSON object from file
```

File: integrations/pagerduty.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+        Parameters
+        ----------
+        file_location : str
+            Path to file alert location.
```
Comment (MarcelKemp) [inline integrations/pagerduty.py:224]:
```suggestion
            Path to the JSON file location.
```

File: integrations/tests/test_pagerduty.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
+# Copyright (C) 2023, Wazuh Inc.
```
Comment (MarcelKemp) [inline integrations/tests/test_pagerduty.py:1]: Check all new or modified files.
```suggestion
# Copyright (C) 2015, Wazuh Inc.
```

**Review by TomasTurina — APPROVED** [review]

**Review by MarcelKemp — APPROVED** [review]

**Pull request merged**